[AMDGPU] Error on non-global pointer with s_prefetch_data

rampitec opened the pull request.

rampitec committed:
[AMDGPU] Error on non-global pointer with s_prefetch_data

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -9929,9 +9929,9 @@ SDValue SITargetLowering::LowerINTRINSIC_VOID(SDValue Op,
-    // For non-global address space preserve the chain and remove the call.
     if (!AMDGPU::isFlatGlobalAddrSpace(cast<MemSDNode>(Op)->getAddressSpace()))
-      return Op.getOperand(0);
+      report_fatal_error("s_prefetch_data only supports global or constant"
+                         " memory");
```
Can check this in the IR verifier

rampitec replied:
Done.

rampitec committed:
Merge branch 'main' into s_prefgetch_data_err

rampitec committed:
Moved to verifier

rampitec committed:
Address review comments

- Using llvm/Support/AMDGPUAddrSpace.h
- Moved isFlatGlobalAddrSpace/isExtendedGlobalAddrSpace there
- Using intrinsic name in the error message

rampitec commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6281,6 +6281,12 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
   }
+  case Intrinsic::amdgcn_s_prefetch_data: {
+    unsigned AS = Call.getArgOperand(0)->getType()->getPointerAddressSpace();
+    Check(AS == 0 || AS == 1 || AS == 4,
```
Done

rampitec commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -6281,6 +6281,12 @@ void Verifier::visitIntrinsicCall(Intrinsic::ID ID, CallBase &Call) {
+  case Intrinsic::amdgcn_s_prefetch_data: {
+    unsigned AS = Call.getArgOperand(0)->getType()->getPointerAddressSpace();
+    Check(AS == 0 || AS == 1 || AS == 4,
+          "s_prefetch_data only supports global or constant memory");
```
Done

rampitec committed:
clang-format
